### PR TITLE
Add method to detect if task is already running

### PR DESCRIPTION
**Pull request by tcollier**
@jrichard @kristjan 

Add `Process#running?` to detect if the rake task (with the given args) is already running. This will be used in scheduled-analytics to act as a lock so that we don't run the same rake task twice, e.g. 

``` ruby
class UpdateTablesJob
  @queue = :low

  def self.perform(*tables)
    process = Dynosaur::Process::Heroku.new(task: 'etl:update_tables', args: tables)
    process.start unless process.running?
  end
end
```

### CODE REVIEW

File: lib/dynosaur/process.rb [hunk 1]
```diff
@@ -3,29 +3,26 @@
...
   class Process
     def initialize(task:, args: [], opts: {})
-      @task = task
-      @args = args
+      @rake_command = Utils::RakeCommand.new(task: task, args: args)
       after_initialize(opts)
     end
 
+    def running?
+      klass = self.class.const_get('Finder')
```
Comment (kristjan) [inline lib/dynosaur/process.rb]: I think you can just `self.class::Finder` here

File: lib/dynosaur/utils/rake_command.rb [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+      def to_s
+        formatted_args = args.map do |arg|
+          arg.is_a?(String) ? arg.shellescape : arg
+        end.join(',')
+        task_with_args = args.empty? ? task : "#{task}[#{formatted_args}]"
+        "rake #{task_with_args} --trace"
+      end
+
+      def ==(other)
+        task == other.task && args.map(&:to_s) == other.args.map(&:to_s)
```
Comment (kristjan) [inline lib/dynosaur/utils/rake_command.rb]: Couldn't this just be `to_s == other.to_s`? Then we have just one non-Ruby representation to worry about.

File: spec/dynosaur/process/heroku/finder_spec.rb [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        let(:dynos) do
+          [{ 'type' => 'run', 'command' => 'bundle exec rake fake:task[foo]' }]
+        end
+
+        it 'returns false' do
+          expect(subject.exists?).to eq(false)
+        end
+      end
+
+      context 'and it is running a namespaed task' do
```
Comment (kristjan) [inline spec/dynosaur/process/heroku/finder_spec.rb]: :book: namespae and neuter your pets!

File: spec/dynosaur/utils/rake_command_spec.rb [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+  describe '==' do
+    context 'when one has a String arg and the other a Fixnum' do
+      subject { Dynosaur::Utils::RakeCommand.new(task: 'fake:task', args: [10]) }
+
+      let(:other) { Dynosaur::Utils::RakeCommand.new(task: 'fake:task', args: ['10']) }
+
+      it 'returns true' do
+        expect(subject).to eq(other)
+      end
```
Comment (kristjan) [inline spec/dynosaur/utils/rake_command_spec.rb:162]: We could probably use a few more specs in here

File: spec/spec_helper.rb [hunk 5]
```diff
@@ -8,3 +8,9 @@
   config.app_name = 'dynosaur-test-app'
   config.api_key = 'dynosaur-test-key'
 end
+
+RSpec.configure do |config|
+  config.mock_with :rspec do |mocks|
+    mocks.verify_partial_doubles = true
+  end
```
Comment (kristjan) [inline spec/spec_helper.rb:15]: :dancers: :heavy_check_mark:

File: lib/dynosaur/utils/os.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+          ps_array.shift # Remove the header row
+          ps_array.map do |command|
+            Struct.new(:command).new(command)
+          end
+        end
+
+        private
+
+        def command_ps
+          `ps -o command`
```
Comment (kristjan) [inline lib/dynosaur/utils/os.rb]: Are there any OS-dependent issues we need to worry about here? We could use something like [sys-proctable](https://github.com/djberg96/sys-proctable)

Comment (tcollier) [inline lib/dynosaur/utils/os.rb]: Good call, I'll update

File: lib/dynosaur/utils/rake_command.rb (hunk 2 above)
Comment (tcollier) [inline lib/dynosaur/utils/rake_command.rb]: That'll work!

File: lib/dynosaur/process.rb (hunk 1 above)
Comment (tcollier) [inline lib/dynosaur/process.rb]: That works.

**Comment by tcollier** [comment]
@kristjan Updated based on your comments, LGTY?

**Comment by kristjan** [comment]
LGTM :ship:

**Pull request merged**